In [1]:
# pip install lyricsgenius

# Import libraries

In [2]:
import pandas as pd
from config import Genius_token
import lyricsgenius
from textblob import TextBlob
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

In [3]:
genius = lyricsgenius.Genius(Genius_token)

# 1. Search for Artists or Songs:

In [4]:
artist_name = "Imagine Dragons" 
artist = genius.search_artist(artist_name, max_songs=5, sort="title")
for song in artist.songs:
    print(f"\nSong: {song.title}")
    print(f"\nLyrics: {song.lyrics[:200]}...")

Searching for songs by Imagine Dragons...

Song 1: "#1"
Song 2: "30 Lives"
Song 3: "30 Lives (Original Demo)"
Song 4: "42"
"5th Chord" is not valid. Skipping.
Song 5: "All Eyes"

Reached user-specified song limit (5).
Done. Found 5 songs.

Song: #1

Lyrics: 24 ContributorsTranslationsРусскийEspañol#1 Lyrics[Verse 1]
Could someone take me home where I'm supposed to be?
I'm on my back again, caught up in something
Could someone take me home? I could use a ...

Song: 30 Lives

Lyrics: 15 Contributors30 Lives Lyrics[Verse 1]
Good morning my love
It's been a while since we have talked
And you have grown
Your eyes are older than the tallest tree you see
Because thirty lives can be ext...

Song: 30 Lives (Original Demo)

Lyrics: 5 Contributors30 Lives (Original Demo) Lyrics[Verse 1]
Good morning, my love
It's been a while since we first talked
And I have grown
My eyes are older than the tallest tree you see
Thirty lives can b...

Song: 42

Lyrics: 3 Contributors42 Lyrics[Verse 1]
Take my Numb

We search for up to 5 songs by the artist "Imagine Dragons" and print the first 200 characters of the lyrics for each song.

# 2. Retrieve Lyrics for Specific Songs:

In [5]:
song = genius.search_song("Unstoppable", "Sia")
print(f"Lyrics for {song.title}:")
print(song.lyrics[::]) 

Searching for "Unstoppable" by Sia...
Done.
Lyrics for Unstoppable:
81 ContributorsTranslationsEspañolPortuguêsFrançaisΕλληνικάالعربيةUnstoppable Lyrics[Verse 1]
I'll smile, I know what it takes to fool this town
I'll do it 'til the sun goes down and all through the night time
Oh, yeah
Oh, yeah, I'll tell you what you wanna hear
Keep my sunglasses on while I shed a tear
It's never the right time, yeah-yeah

[Pre-Chorus]
I put my armor on, show you how strong I am
I put my armor on, I'll show you that I am

[Chorus]
I'm unstoppable, I'm a Porsche with no brakes
I'm invincible, yeah, I win every single game
I'm so powerful, I don't need batteries to play
I'm so confident, yeah, I'm unstoppable today
Unstoppable today, unstoppable today
Unstoppable today, I'm unstoppable today

[Verse 2]
Breakdown, only alone I will cry out loud
You'll never see what's hiding out
Hiding out deep down, yeah-yeah
I know, I've heard that to let your feelings show
Is the only way to make friendships grow
But 

# 3. Retrieve Multiple Songs:

In [7]:
artists = ["Linkin Park", "Ed Sheeran", "Imagine Dragons", "The Weeknd", "Drake"]
songs = []
for artist_name in artists:
    artist = genius.search_artist(artist_name, max_songs=3, sort="title")
    for song in artist.songs:
        songs.append(song)
        print(f"\nSong: {song.title} by {artist_name}")
        print(f"\nLyrics: {song.lyrics[:300]}...\n")

Searching for songs by Linkin Park...

Song 1: "1Stp Klosr"
"21 Stitches (Unreleased 2006 “Given Up” Demo)" is not valid. Skipping.
Song 2: "26 Lettaz in Da Alphabet"
Song 3: "A.06"

Reached user-specified song limit (3).
Done. Found 3 songs.

Song: 1Stp Klosr by Linkin Park

Lyrics: 16 Contributors1Stp Klosr Lyrics[Intro: Chester Bennington]
…Room to breathe
'Bout to…
Break
Breathe
I'm about to break
Breathe
I need room to breathe
Breathe
I need room to breathe
Breathe
I need room to breathe

[Verse 1: Chester Bennington]
I cannot take this anymore
Saying everything I've said b...


Song: 26 Lettaz in Da Alphabet by Linkin Park

Lyrics: 10 Contributors26 Lettaz in Da Alphabet Lyrics[Intro]
Man, the other day
I was trying to figure out how many letters was in the alphabet
It took me all day, but you know I did it
You wanna know how many letters there are?
26 letters in the alphabet!

[Verse 1]
26 letters in the alphabet
I wear a sho...


Song: A.06 by Linkin Park

Lyrics: 5 Contributor

# 4. Sentiment Analysis:

In [8]:
def get_sentiment(lyrics):
    blob = TextBlob(lyrics)
    sentiment = blob.sentiment.polarity
    return sentiment

In [9]:
song_data = []

artists = ["Linkin Park", "Ed Sheeran", "Imagine Dragons", "The Weeknd", "Drake"]
for artist_name in artists:
    artist = genius.search_artist(artist_name, max_songs=3, sort="title")
    for song in artist.songs:
        sentiment_score = get_sentiment(song.lyrics)

        song_data.append([song.title, artist_name, round(sentiment_score, 4)])

df = pd.DataFrame(song_data, columns=['Title', 'Artist', 'Sentiment Score'])
df

Searching for songs by Linkin Park...

Song 1: "1Stp Klosr"
"21 Stitches (Unreleased 2006 “Given Up” Demo)" is not valid. Skipping.
Song 2: "26 Lettaz in Da Alphabet"
Song 3: "A.06"

Reached user-specified song limit (3).
Done. Found 3 songs.
Searching for songs by Ed Sheeran...

Song 1: "1000 Nights"
Song 2: "2step"
Song 3: "2step (American Remix)"

Reached user-specified song limit (3).
Done. Found 3 songs.
Searching for songs by Imagine Dragons...

Song 1: "#1"
Song 2: "30 Lives"
Song 3: "30 Lives (Original Demo)"

Reached user-specified song limit (3).
Done. Found 3 songs.
Searching for songs by The Weeknd...

Song 1: "6 Inch Heel"
Song 2: "Acquainted"
Song 3: "Adaptation"

Reached user-specified song limit (3).
Done. Found 3 songs.
Searching for songs by Drake...

Song 1: "0 to 100 / The Catch Up"
Song 2: "100 GB (Un-Official)"
Song 3: "10 Bands"

Reached user-specified song limit (3).
Done. Found 3 songs.


Title           Artist  Sentiment Score
0                 1Stp Klosr      Linkin Park          -0.1962
1   26 Lettaz in Da Alphabet      Linkin Park           0.0450
2                       A.06      Linkin Park           0.0000
3                1000 Nights       Ed Sheeran           0.0058
4                      2step       Ed Sheeran           0.1206
5     2step (American Remix)       Ed Sheeran           0.1612
6                         #1  Imagine Dragons           0.1171
7                   30 Lives  Imagine Dragons          -0.0627
8   30 Lives (Original Demo)  Imagine Dragons          -0.0181
9                6 Inch Heel       The Weeknd           0.1160
10                Acquainted       The Weeknd           0.1510
11                Adaptation       The Weeknd           0.1434
12   0 to 100 / The Catch Up            Drake           0.0357
13      100 GB (Un-Official)            Drake           0.0000
14                  10 Bands            Drake          -0.0654

## Using VADER Sentiment Analysis:

In [10]:
analyzer = SentimentIntensityAnalyzer()

# Define the function to calculate sentiment score using VADER
def get_vader_sentiment(lyrics):
    sentiment = analyzer.polarity_scores(lyrics)
    return sentiment['compound']  # Compound score ranges from -1 (negative) to +1 (positive)

# Initialize an empty list to store the song details
song_data = []

artists = ["Linkin Park", "Ed Sheeran", "Imagine Dragons", "The Weeknd", "Drake"]
for artist_name in artists:
    artist = genius.search_artist(artist_name, max_songs=3, sort="title")
    for song in artist.songs:
        sentiment_score = get_vader_sentiment(song.lyrics)

        song_data.append([song.title, artist_name, round(sentiment_score, 4)])

# Create a DataFrame from the list
df_vader = pd.DataFrame(song_data, columns=['Title', 'Artist', 'Sentiment Score'])
df_vader


Searching for songs by Linkin Park...

Song 1: "1Stp Klosr"
"21 Stitches (Unreleased 2006 “Given Up” Demo)" is not valid. Skipping.
Song 2: "26 Lettaz in Da Alphabet"
Song 3: "A.06"

Reached user-specified song limit (3).
Done. Found 3 songs.
Searching for songs by Ed Sheeran...

Song 1: "1000 Nights"
Song 2: "2step"
Song 3: "2step (American Remix)"

Reached user-specified song limit (3).
Done. Found 3 songs.
Searching for songs by Imagine Dragons...

Song 1: "#1"
Song 2: "30 Lives"
Song 3: "30 Lives (Original Demo)"

Reached user-specified song limit (3).
Done. Found 3 songs.
Searching for songs by The Weeknd...

Song 1: "6 Inch Heel"
Song 2: "Acquainted"
Song 3: "Adaptation"

Reached user-specified song limit (3).
Done. Found 3 songs.
Searching for songs by Drake...

Song 1: "0 to 100 / The Catch Up"
Song 2: "100 GB (Un-Official)"
Song 3: "10 Bands"

Reached user-specified song limit (3).
Done. Found 3 songs.


Title           Artist  Sentiment Score
0                 1Stp Klosr      Linkin Park           0.5948
1   26 Lettaz in Da Alphabet      Linkin Park          -0.5940
2                       A.06      Linkin Park           0.0000
3                1000 Nights       Ed Sheeran           0.9662
4                      2step       Ed Sheeran           0.9945
5     2step (American Remix)       Ed Sheeran           0.9955
6                         #1  Imagine Dragons           0.9809
7                   30 Lives  Imagine Dragons           0.9655
8   30 Lives (Original Demo)  Imagine Dragons           0.9618
9                6 Inch Heel       The Weeknd          -0.9871
10                Acquainted       The Weeknd           0.9736
11                Adaptation       The Weeknd           0.9947
12   0 to 100 / The Catch Up            Drake          -0.9988
13      100 GB (Un-Official)            Drake          -0.2732
14                  10 Bands            Drake           0.9531

# Insights from Comparing the Two Sentiment Analysis Methods:
**Sentiment Range**:

- TextBlob (get_sentiment): The sentiment scores are in the range from -1 (negative) to +1 (positive), with scores generally close to 0 for more neutral tones (e.g., songs like "A.06" by Linkin Park and "100 GB (Un-Official)" by Drake, which have a score of 0.0).
- VADER (get_vader_sentiment): The VADER scores also range from -1 to +1, but the results show a wider spread. For example, VADER assigns high positive sentiment to songs like "2step" by Ed Sheeran with a score of 0.9945, but also gives negative scores like -0.9871 for "6 Inch Heel" by The Weeknd.

**Sensitivity to Song Lyrics**:

- TextBlob seems to provide more moderate sentiment scores, likely due to its simpler method of analysis. It often provides near-neutral scores (around 0), even for songs with relatively emotional content.
- VADER appears to be more sensitive to emotion in lyrics. It detects strong positive or negative sentiment more distinctly. For instance, "2step (American Remix)" by Ed Sheeran has a score of 0.9955 in VADER, indicating a strongly positive tone, whereas TextBlob assigns it only a modest 0.1206.

**Contrast in Song Sentiment**:

- Songs that both methods identify as positive include "2step" by Ed Sheeran, where VADER gives a very high score (0.9945) compared to TextBlob’s 0.1206. This could indicate that VADER is better at picking up on enthusiastic or positive sentiment in songs with energetic or upbeat tones.
- On the other hand, songs like "6 Inch Heel" by The Weeknd show a stark contrast: VADER assigns a strongly negative sentiment score (-0.9871), while TextBlob gives a mild negative score (-0.1962). This could be because VADER is more sensitive to intense negative emotion in the lyrics.

**Neutral Sentiment**:

- Both methods agree on neutral or almost neutral sentiment for songs like "A.06" by Linkin Park, which gets a score of 0.0 from both methods. This suggests that these methods are aligned when the sentiment is subtle or unclear.

**Genre and Artist Influence**:

- The differences in sentiment analysis scores might reflect the artists' emotional range or the lyrical content typical of their genre. For example, Drake's songs like "0 to 100 / The Catch Up" are strongly negative in VADER (-0.9987), which might be due to the intense, raw emotion often found in his lyrics. TextBlob still shows negative sentiment for these songs but to a lesser degree.
# Conclusion:
- **VADER** is generally more sensitive to extreme positive and negative emotions, making it suitable for analyzing songs with clear emotional extremes.
- **TextBlob** provides more moderate sentiment scores, making it more suitable for capturing nuanced or subtle feelings.